text/0000-cargo-test-coverage.md

```diff
@@ -0,0 +1,366 @@
+This means more boiler plate code for each Rust developer that simply wants to profile their own crate.
+I believe the feature this RFC proposes would both be a cleaner solution long term and more in line
+with the Cargo workflow of potentially reading these kinds of behaviors from the `Cargo.toml` manifest
+file.
+
+### Alternative 4: use existing custom subcommands to run code coverage analysis
+
+Supporting this alternative would mean that there wouldn't need to be any changes to Cargo
+at all.
+
+There are multiple custom subcommands that exist today to achieve instrumenting Rust libraries
+analyzing code coverage metrics. A couple of examples would be cargo-llvm-cov and cargo-tarpaulin.
+Both of these custom subcommands are available via crates.io and support running tests with code
+coverage enabled and creating a coverage report to be viewed by a Rust developer.
+
+Cargo-llvm-cov currently leverages the `cargo-config` subcommand which is still unstable. To do so,
+cargo-llvm-cov sets the `RUSTC_BOOTSTRAP` environment variable to allow its usage from a stable
+toolchain. This is not a recommended approach especially for Rust developers that want to use
+such tools in production code.
+
+Tarpaulin has a great set of features for collecting and analyzing coverage metrics but it only
+supports x86_64 processors running Linux which limits how this can be used by Rust developers
+working on other platforms such as Windows.
+
+# Prior art
+[prior-art]: #prior-art
+
+## VSInstr
+
+Visual Studio ships with a tool `vsinstr.exe` which has support for instrumenting binaries after
+they have already been built. Since LLVMs instrumentation-based code coverage hooks into each object
+file it generates this scenario is a bit different than the feature this RFC proposes. `vsinstr` does
+allow for excluding namespaces of functions to skip over so that everything within a binary does not
+get instrumented.
+
+ - [vsinstr](https://docs.microsoft.com/en-us/previous-versions/visualstudio/visual-studio-2017/profiling/vsinstr?view=vs-2017)
+
+## gcov based coverage
+
+Rust also has support for another type of code coverage, a GCC-compatible, gcov-based coverage implementation.
+This can be enabled through the `-Z profile` flag. This uses debuginfo to derive code coverage. This is different
+than the source-based code coverage which allows for a more precise instrumentation to be done.
+
+ - [Source-Based Code Coverage](https://blog.rust-lang.org/inside-rust/2020/11/12/source-based-code-coverage.html)
+
+# Unresolved questions
+[unresolved-questions]: #unresolved-questions
+
+- Are there any drawbacks from having Cargo set the `LLVM_PROFILE_FILE` environment variable that LLVM uses to name each
+of the generated `profraw` files? This is used to ensure each test run generates unique profiling data as opposed to overwriting
+the previous run.
+
+# Future possibilities
```

Comment:
It might be worth mentioning here some kinds of tests that are supported by cargo-llvm-cov and do not seem to be supported by this RFC.

- other (builtin) subcommands (e.g., `cargo run`)
- merge coverage of different runs
- coverage of C/C++ code linked to Rust lib/bin

cargo-llvm-cov also supports the following, but I think they are out of the scope of cargo.

- [external tests](https://github.com/taiki-e/cargo-llvm-cov#get-coverage-of-external-tests)
- third-party test libraries (e.g., trybuild)
- other (third-party) subcommands (e.g., `cargo nextest`)